sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -0,0 +1,406 @@
+        return new JdbcOptions(dataSource, query, username, password);
+      }
+
+      public JdbcOptions withQuery(String query) {
+        return new JdbcOptions(dataSource, query, username, password);
+      }
+
+      public JdbcOptions withRowMapper(RowMapper rowMapper) {
+        return new JdbcOptions(dataSource, query, username, password);
+      }
+
+      public JdbcOptions withUsername(String username) {
+        return new JdbcOptions(dataSource, query, username, password);
+      }
+
+      public JdbcOptions withPassword(String password) {
+        return new JdbcOptions(dataSource, query, username, password);
+      }
+
+      public void validate() {
+        Preconditions.checkNotNull(dataSource, "dataSource");
+        Preconditions.checkNotNull(query, "query");
+      }
+
+      public void populateDisplayData(DisplayData.Builder builder) {
+        builder.add(DisplayData.item("dataSource", dataSource.getClass().getName()));
+        builder.add(DisplayData.item("query", query));
+        builder.addIfNotNull(DisplayData.item("username", username));
+      }
+
+      public DataSource getDataSource() {
+        return dataSource;
+      }
+
+      public String getQuery() {
+        return query;
+      }
+
+      @Nullable
+      public String getUsername() {
+        return username;
+      }
+
+      @Nullable
+      public String getPassword() {
+        return password;
+      }
+    }
+
+    /**
+     * A {@link DoFn} executing the SQL query to read from the database.
+     */
+    public static class ReadFn<T> extends DoFn<JdbcOptions, T> {
+
+      private final RowMapper<T> rowMapper;
+
+      private ReadFn(RowMapper<T> rowMapper) {
+        this.rowMapper = rowMapper;
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext context) throws Exception {
+        JdbcOptions options = context.element();
+
+        try (Connection connection = (options.getUsername() != null)
+            ? options.getDataSource().getConnection(options.getUsername(), options.getPassword())
+            : options.getDataSource().getConnection()) {
+
+          try (PreparedStatement statement = connection.prepareStatement(options.getQuery())) {
+            try (ResultSet resultSet = statement.executeQuery()) {
+              while (resultSet.next()) {
+                T record = rowMapper.mapRow(resultSet);
+                if (record != null) {
+                  context.output(record);
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * An interface used by the JdbcIO Write for mapping {@link PCollection} elements as a  rows of a
+   * ResultSet on a per-row basis.
+   * Implementations of this interface perform the actual work of mapping each row to a result
+   * object used in the {@link PCollection}.
+   */
+  public interface ElementInserter<T> extends Serializable {
+
+    PreparedStatement insert(T element, Connection connection);
```

Comment:
Perhaps it'd be better to make the API only slightly less generic but slightly more clear to use / harder to mis-use, by making the following change:
- Pass the SQL string to the Write transform
- Make the WriteFn create the prepared statement itself, in the Setup method
- Make ElementInserter's signature be: void setParameters(T element, PreparedStatement statement) (and rename it to PreparedStatementSetter?)

This way:
- the PreparedStatement object can be reused, which will give much better performance
- the user doesn't have to call .prepareStatement
- the user can not accidentally mess up the connection by doing something to it that the WriteFn doesn't expect (e.g. calling commit, or close, or doing queries)
- it is easier to understand the semantics of the transform because it always uses the same SQL just with different parameters
- an ElementInserter is easier to test without a Connection
- it becomes easier to write composable ElementInserter classes, e.g. you could imagine expressing something like { st.setString(1, t.getKey()); st.setInteger(2, t.getValue()); } as "allOf(stringAt(1, key()), integerAt(2, value()))" where allOf, stringAt, integerAt, key, value are static utility functions. Not suggesting to do this in the current PR, but I once developed a library of composable row mappers and prepared statement setters like this at one of my previous companies and it was very handy to use.